src/test/gcp/cloudfunctions.spec.ts

```diff
@@ -259,82 +261,112 @@ describe("cloudfunctions", () => {
-        accounts.push(`serviceAccount-${i}@`);
-      }
-
-      const policy = cloudfunctions.generateIamPolicy(project, accounts);
-
-      expect(policy.bindings.length).to.eq(1);
-      for (let i = 0; i < 10; i++) {
-        expect(policy.bindings[0].members.length).to.eq(10);
-        expect(policy.bindings[0].members[i]).to.eq(
-          `serviceAccount:serviceAccount-${i}@${project}.iam.gserviceaccount.com`
-        );
-      }
+    it("should reject if the setting the IAM policy fails", async () => {
+      apiRequestStub.onFirstCall().resolves({});
+      apiRequestStub.onSecondCall().throws("Error calling set api.");
+
+      await expect(cloudfunctions.setInvoker("project", "function", ["public"])).to.be.rejectedWith(
+        "Failed to set the IAM Policy on the function function"
+      );
+      expect(apiRequestStub).to.be.calledTwice;
     });
 
-    it("should generate a policy for a single full service account", () => {
-      const serviceAccount = "service-account@project.iam.gserviceaccount.com";
-      const project = "project";
+    it("should set a basic policy on a function without any polices", async () => {
+      apiRequestStub.onFirstCall().resolves({});
+      apiRequestStub.onSecondCall().callsFake((method: any, resource: any, options: any) => {
+        expect(options.data.policy).to.deep.eq({
+          bindings: [
+            {
+              role: "roles/cloudfunctions.invoker",
+              members: ["allUsers"],
+            },
+          ],
+          etag: "",
+          version: 3,
+        });
 
-      const policy = cloudfunctions.generateIamPolicy(project, serviceAccount);
+        return Promise.resolve();
+      });
 
-      expect(policy.bindings.length).to.eq(1);
-      expect(policy.bindings[0].members.length).to.eq(1);
-      expect(policy.bindings[0].members[0]).to.eq(`serviceAccount:${serviceAccount}`);
+      await expect(cloudfunctions.setInvoker("project", "function", ["public"])).to.not.be.rejected;
+      expect(apiRequestStub).to.be.calledTwice;
     });
 
-    it("should generate a policy for multiple full service accounts", () => {
-      const project = "project";
-      const accounts = [];
-      for (let i = 0; i < 10; i++) {
-        accounts.push(`serviceAccount-${i}@project.iam.gserviceaccount.com`);
-      }
+    it("should set the policy with private invoker with active policies", async () => {
+      apiRequestStub.onFirstCall().resolves({
+        bindings: [
+          { role: "random-role", members: ["user:pineapple"] },
+          { role: "roles/cloudfunctions.invoker", members: ["some-service-account"] },
+        ],
+        etag: "1234",
+        version: 3,
+      });
+      apiRequestStub.onSecondCall().callsFake((method: any, resource: any, options: any) => {
+        expect(options.data.policy).to.deep.eq({
+          bindings: [
+            { role: "random-role", members: ["user:pineapple"] },
+            { role: "roles/cloudfunctions.invoker", members: [] },
```

Comment:
Interesting. It didn't even occur to me to have an empty binding instead of just removing the binding. Does this work?

Reply:
From my testing it does, sending an empty array to the api will remove the binding on a subsequent GET request.